File: ai/mcts.py
import numpy as np
import math

class MCTSNode:
    def __init__(self, state, parent=None, prior=0):
        self.state = state
        self.parent = parent
        self.children = {} # move -> MCTSNode
        self.visit_count = 0
        self.value_sum = 0
        self.prior = prior
        self.state_hash = None # To be used with C# engine

    @property
    def value(self):
        if self.visit_count == 0:
            return 0
        return self.value_sum / self.visit_count

    def select_child(self, c_puct):
        best_score = -float('inf')
        best_move = -1
        best_child = None

        for move, child in self.children.items():
            score = child.value + c_puct * child.prior * math.sqrt(self.visit_count) / (1 + child.visit_count)
            if score > best_score:
                best_score = score
                best_move = move
                best_child = child
        
        return best_move, best_child

    def expand(self, action_probs):
        for move, prob in action_probs:
            if move not in self.children:
                self.children[move] = MCTSNode(None, parent=self, prior=prob)

    def update(self, value):
        self.visit_count += 1
        self.value_sum += value
        if self.parent:
            self.parent.update(-value) # Negate for opponent

File: ChessRL.Engine/MCTS.cs
using System.Collections.Concurrent;

namespace ChessRL.Engine;

public class MCTSNode
{
    public Board State { get; }
    public Move Move { get; }
    public MCTSNode Parent { get; }
    public List<MCTSNode> Children { get; }
    
    public int VisitCount { get; set; }
    public float ValueSum { get; set; }
    public float Prior { get; }
    
    public bool IsExpanded => Children.Count > 0;

    public MCTSNode(Board state, Move move = default, MCTSNode parent = null, float prior = 0)
    {
        State = state;
        Move = move;
        Parent = parent;
        Prior = prior;
        Children = new List<MCTSNode>();
    }

    public float GetValue() => VisitCount == 0 ? 0 : ValueSum / VisitCount;

    public MCTSNode Select(float cpuct)
    {
        MCTSNode bestChild = null;
        double bestScore = double.NegativeInfinity;

        foreach (var child in Children)
        {
            // PUCT Formula: Q(s,a) + U(s,a)
            double q = child.GetValue();
            double u = cpuct * child.Prior * Math.Sqrt(VisitCount) / (1 + child.VisitCount);
            double score = q + u;

            if (score > bestScore)
            {
                bestScore = score;
                bestChild = child;
            }
        }

        return bestChild;
    }

    public void Expand(Span<Move> moves, float[] priors)
    {
        for (int i = 0; i < moves.Length; i++)
        {
            // Apply the move to create a NEW board state for the child
            Board nextState = State.Clone();
            nextState.MakeMove(moves[i]);
            Children.Add(new MCTSNode(nextState, moves[i], this, priors[i]));
        }
    }

    public void Backpropagate(float value)
    {
        MCTSNode current = this;
        float v = value;
        while (current != null)
        {
            current.VisitCount++;
            current.ValueSum += v;
            v = -v; // Flip value for opponent
            current = current.Parent;
        }
    }
}
